src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
 
+static void nxt_array_elts_copy(void **dst_elts, nxt_array_t *src_array);
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
 nxt_array_t *
-nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
+nxt_array_create(nxt_mp_t *mp, nxt_uint_t arr_size, size_t elt_size)
 {
-    nxt_array_t  *array;
+    nxt_uint_t     nalloc;
+    nxt_array_t    *array;
+    void           *elts;
 
-    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);
+    array = nxt_mp_alloc(mp, sizeof(nxt_array_t));
 
     if (nxt_slow_path(array == NULL)) {
         return NULL;
     }
 
-    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
+    nalloc = nxt_array_calc_alloc(arr_size);
+    elts = nxt_mp_alloc(mp, elt_size * nalloc);
+
+    if (nxt_slow_path(elts == NULL)) {
+        return NULL;
+    }
+
+    array->elts = elts;
     array->nelts = 0;
-    array->size = size;
-    array->nalloc = n;
+    array->size = elt_size;
+    array->nalloc = nalloc;
     array->mem_pool = mp;
 
     return array;
 }
 
 
+/*
+ * nxt_array_destroy - Destroy the existing array.
+ */
 void
 nxt_array_destroy(nxt_array_t *array)
 {
-    if (array->elts != nxt_pointer_to(array, sizeof(nxt_array_t))) {
-        nxt_mp_free(array->mem_pool, array->elts);
-    }
+    if (nxt_fast_path(array != NULL)) {
+
+        array->nelts = 0;
 
-    nxt_mp_free(array->mem_pool, array);
+        (void) nxt_mp_free(array->mem_pool, array->elts);
+        (void) nxt_mp_free(array->mem_pool, array);
+
+    }
 }
 
 
+/*
+ * nxt_array_add - Add an element to the existing array.
+ */
 void *
 nxt_array_add(nxt_array_t *array)
 {
-    void      *p;
-    uint32_t  nalloc, new_alloc;
-
-    nalloc = array->nalloc;
+    if (nxt_fast_path(array != NULL)) {
 
-    if (array->nelts == nalloc) {
+        nxt_uint_t    nalloc;
+        void          *elt, *elts;
 
-        if (nalloc < 16) {
```

Comment:
- fixed the formula for calculating the required amount of memory (nalloc) allocated for the created array;

it is better to use an even value (2, 4, 8, 16, 32, etc) and also a multiple of 8.

Reply:
What is this? One more?

**Please provide 2 common cases:**
one for `x < 16 = x * 2`
one for `x > 16 = x + x / 2`

Otherwise, it is not clear to me what specific cases you are talking about and how I can get acquainted with them, check and make sure that the information you provided is objective, and not fictitious. 

How is the result `16 * 2 = 32` worse than `15 * 2 = 30`?

How then does your overgrowth logic fit into the initial initialization of an empty array in size `(nalloc == 0 ) ? 4`?

Reply:
In most cases the initial array size is set to fit all the values. But this value can be not very precise and sometimes the array can grow a bit. The idea is to find a balance between too much memory wasted and too many reallocations. Current logic avoids multiple reallocations for small values by using more steep growing formula for low number of elements and at the same time it avoids too much overallocation for adding a few more elements over already relatively high number of elements.

4 and 16 are constants which obtained empirically by Igor Sysoev based on his 20+ years of experience in nginx development.

Reply:
> In most cases the initial array size is set to fit all the values. But this value can be not very precise and sometimes the array can grow a bit. The idea is to find a balance between too much memory wasted and too many reallocations. Current logic avoids multiple reallocations for small values by using more steep growing formula for low number of elements and at the same time it avoids too much overallocation for adding a few more elements over already relatively high number of elements.
> 
> 4 and 16 are constants which obtained empirically by Igor Sysoev based on his 20+ years of experience in nginx development.

I accept the empirical method of **Igor Sysoev**, since I have great respect for the work he has done, so let's leave it as it is. 

The actual number of uses of the value 4 (nalloc) with the initial 0 is unknown to me at the moment. Perhaps in the future you should evaluate the real optimal value for these parameters, since most scenarios assume a known value in advance, which in most cases is greater than 0. 

In this regard, I think this will not be any problem for you if this issue really affects performance, then there must be a real basis and based on performance testing of this logic so that the values ​​bare justified and you do not have to refer first to existing use cases and then to an empirical method.

**I'm starting from the logic:**
Don't allocate too much memory unless you really need to. When is the next call to reallocate memory for the array being expanded? - unknown, possibly never. In that case, which is more important? Allocated and never used memory, or one extra call to memory when you need to expand the array?

Reply:
4 was just a good start for the *2 rule... assuming that most elements are just 2-4 pointers long and nxt_mp_alloc() won't work good for very small allocations anyway.

> which is more important? Allocated and never used memory, or one extra call to memory when you need to expand the array?

Sure it depends on how big is the extra memory that is allocated and never used. Any extra call that can possibly lead to a syscall is very bad. Too much allocation isn't good either.

Reply:
Note also, that nxt_mp_alloc() doesn't always allocate the size that you've asked for. Sometimes it over-allocates too.